### PR TITLE
refactor(v3): add role slice (DEV-1010)

### PR DESCRIPTION
**Pull request by mpro7**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: DEV-1010

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+               name2,
+               description2,
+               users2,
+               permission2
+             )
+
+        roles <- handler.getRoles()
+      } yield assertTrue(roles.size == 2)
+    },
+    test("store a role and update its name") {
```
Comment (BalduinLandolt) [inline dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:66]: all the "update" tests could be in a sub-spec for updating. might make the test results a bit more readable

File: dsp-role/repo/src/main/scala/dsp/role/repo/impl/RoleRepoLive.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  override def storeRole(r: Role): UIO[RoleId] =
+    (for {
+      _ <- roles.put(r.id.uuid, r)
+    } yield r.id).commit.tap(_ => ZIO.logInfo(s"Stored role with ID: ${r.id.uuid}"))
+
+  /**
+   * @inheritDoc
+   */
+  override def getRoles(): UIO[List[Role]] =
+    roles.values.commit.tap(rolesList => ZIO.logInfo(s"Found roles number: ${rolesList.size}"))
```
Comment (BalduinLandolt) [inline dsp-role/repo/src/main/scala/dsp/role/repo/impl/RoleRepoLive.scala:37]:
```suggestion
    roles.values.commit.tap(rolesList => ZIO.logInfo(s"Found number of roles: ${rolesList.size}"))
```
might read a bit more elegantly... idk

File: dsp-role/repo/src/main/scala/dsp/role/repo/impl/RoleRepoLive.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * @inheritDoc
+   */
+  override def getRoleById(id: RoleId): IO[Option[Nothing], Role] =
+    roles
+      .get(id.uuid)
+      .commit
+      .some
+      .tapBoth(
+        _ => ZIO.logInfo(s"Role with ID: ${id.uuid} not found"),
```
Comment (BalduinLandolt) [inline dsp-role/repo/src/main/scala/dsp/role/repo/impl/RoleRepoLive.scala:48]:
```suggestion
        _ => ZIO.logInfo(s"Role with ID '${id.uuid}' not found"),
```
I don't think the colon in the middle of the word works nicely... maybe use single quotes instead?

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  val roleTests =
+    test("store two roles and retrieve them all") {
+      for {
+        role1 <- role1.toZIO
+        role2 <- role2.toZIO
+        _     <- RoleRepo.storeRole(role1)
+        _     <- RoleRepo.storeRole(role2)
+        roles <- RoleRepo.getRoles()
+      } yield assertTrue(roles.size == 2)
+    } +
```
Comment (BalduinLandolt) [inline dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala:35]: is there a reason for combining those with plus, which leads to increasing indentation?

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoMock.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  override def storeRole(r: Role): UIO[RoleId] =
+    (for {
+      _ <- roles.put(r.id.uuid, r)
+    } yield r.id).commit.tap(_ => ZIO.logInfo(s"Stored role with ID: ${r.id.uuid}"))
+
+  /**
+   * @inheritDoc
+   */
+  override def getRoles(): UIO[List[Role]] =
+    roles.values.commit.tap(rolesList => ZIO.logInfo(s"Found roles number: ${rolesList.size}"))
```
Comment (BalduinLandolt) [inline dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoMock.scala:37]: this is practically the same as RoleRepoLive, so same comments apply here

File: dsp-shared/src/main/scala/dsp/valueobjects/Id.scala [hunk 6]
```diff
@@ -13,6 +13,54 @@ import java.util.UUID
...
+    private val roleIriPrefix = "http://rdfh.ch/roles/"
+
+    /**
+     * Generates a RoleId instance with a UUID and a given IRI.
+     *
+     * @param iri the role's IRI used to create RoleId
+     * @return new RoleId instance
+     */
+    def fromIri(iri: Iri.RoleIri): Validation[Throwable, RoleId] = {
+      val uuid: UUID = UUID.fromString(iri.value.split("/").last)
```
Comment (BalduinLandolt) [inline dsp-shared/src/main/scala/dsp/valueobjects/Id.scala:37]: in my PR I wrapped this in a Try because in theory these methods could throw... but I'm not sure they ever will, so maybe it's also overkill.

**Review by BalduinLandolt — APPROVED** [review]
LGTM, I think I have mostly cosmetics to criticise

File: dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright © 2021 - 2022 Data and Service Center for the Humanities and/or DaSCH Service Platform contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package dsp.role.api
+
+/**
+ * To be implemented...
+ */
+object placeholder {}
```
Comment (irinaschubert) [inline dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala:11]: Is this file needed? Are you going to implement it? I don't think we need to test the repo api (it's just an interface). I think it makes more sense to test the implementations of it.

**Review by subotic — APPROVED** [review]

File: dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala [hunk 8]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright © 2021 - 2022 Data and Service Center for the Humanities and/or DaSCH Service Platform contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package dsp.role.domain
```
Comment (irinaschubert) [inline dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala:6]: I think the package should be `dsp.role.sharedtestdata`.

File: dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import dsp.valueobjects.Role._
+
+/**
+ * Contains shared role test data.
+ */
+object RoleTestData {
+  val id1          = Id.RoleId.make()
+  val name1        = LangString.make("Name", "en")
+  val description1 = LangString.make("Description", "en")
+  val users1       = List(RoleUser(Id.UserId.make().fold(e => throw e.head, v => v)))
```
Comment (irinaschubert) [inline dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala:19]: As it is a list of users, I would add at least 2 users to it.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * The role handler.
+ *
+ * @param repo  the role repository
+ */
+final case class RoleHandler(repo: RoleRepo) {
+
+  /**
+   * Retrieves all roles (sorted by IRI).
```
Comment (irinaschubert) [inline dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:27]: Return values/types are missing in all docstrings in this file.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+      storedRole <- repo.storeRole(role)
+    } yield storedRole).tap(id => ZIO.logInfo(s"Created role with ID: $id"))
+
+  /**
+   * Updates the name of the role.
+   *
+   *  @param id  the roles's ID
+   *  @param newValue  the new name value
+   */
+  def updateName(id: RoleId, newValue: LangString): IO[RequestRejectedException, RoleId] =
```
Comment (irinaschubert) [inline dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:74]: I am not sure if I like this `newValue` pattern. I know that we have it also in the user slice. But if the value is a list, it looks weird. So, maybe we should discuss this.

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+import dsp.role.repo.impl.RoleRepoMock
+import dsp.valueobjects.Id.RoleId
+import dsp.valueobjects.Permission
+import dsp.valueobjects.Role._
+import zio.ZIO
+import zio._
+import zio.test._
+
+/**
+ * This soec is used to test [[RoleHandler]]
```
Comment (irinaschubert) [inline dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:23]:
```suggestion
 * This spec is used to test [[RoleHandler]]
```

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+ */
+object RoleHandlerSpec extends ZIOSpecDefault {
+  def spec = (getRolesTest)
+
+  private val getRolesTest = suite("getRoles")(
```
Comment (irinaschubert) [inline dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:28]: The name of the suite is not exactly correct, it should be something more generic because you are testing not only "get" operations but also updates. I agree with Balduin though that it would be nice to have multiple suites instead of one big.

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+        updatedRoleId <- handler.updatePermission(roleId, newValue)
+        retrievedRole <- handler.getRoleById(updatedRoleId)
+      } yield assertTrue(retrievedRole.name == name1) &&
+        assertTrue(retrievedRole.description == description1) &&
+        assertTrue(retrievedRole.users == users1) &&
+        assertTrue(retrievedRole.permission == newValue)
+    }
+  ).provide(RoleRepoMock.layer, RoleHandler.layer)
+}
```
Comment (irinaschubert) [inline dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:170]: I'm not sure but should there be tests to delete things? Delete a user from the role, delete the description of the role, or even delete a role?

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+          role1           <- role1.toZIO
+          roleId          <- RoleRepo.storeRole(role1)
+          idOfDeletedRole <- RoleRepo.deleteRole(roleId)
+          isIdDeleted     <- RoleRepo.getRoleById(idOfDeletedRole).exit
+        } yield assertTrue(roleId == role1.id) &&
+          assert(isIdDeleted)(fails(equalTo(None)))
+      }
+
+  val roleRepoMockTests = suite("RoleRepoMock")(
+    roleTests
```
Comment (irinaschubert) [inline dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala:54]: I assume this `roleTest` is here on purpose (as placeholder)? If not, it would be wrong and should maybe be deleted (or real tests for the mock repo added).

File: dsp-shared/src/main/scala/dsp/valueobjects/Iri.scala [hunk 16]
```diff
@@ -167,6 +195,8 @@ object IriErrorMessages {
   val ListIriInvalid     = "List IRI is invalid"
   val ProjectIriMissing  = "Project IRI cannot be empty."
   val ProjectIriInvalid  = "Project IRI is invalid."
+  val RoleIriMissing     = "Role IRI cannot be empty."
+  val RoleIriInvalid     = "Role IRI is invalid."
```
[the remark this answers is not in the document]
Comment (irinaschubert) [reply dsp-shared/src/main/scala/dsp/valueobjects/Iri.scala:199]: I agree. This should be done for all the IRIs.

File: dsp-shared/src/main/scala/dsp/valueobjects/Role.scala [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+import dsp.errors.BadRequestException
+import zio.prelude.Validation
+
+object Role {
+
+  /**
+   * LangString value object.
+   *
+   * @param value the [[String]] the value to be validated
+   * @param isoCode thelanguage ISO code to be validated
```
Comment (irinaschubert) [inline dsp-shared/src/main/scala/dsp/valueobjects/Role.scala:17]:
```suggestion
   * @param isoCode the language ISO code to be validated
```

File: dsp-shared/src/main/scala/dsp/valueobjects/V2.scala [hunk 18]
```diff
@@ -116,6 +122,14 @@ object V2IriValidation {
   def isKnoraListIriStr(iri: IRI): Boolean =
     isIri(iri) && iri.startsWith("http://" + IriDomain + "/lists/")
 
+  /**
+   * Returns `true` if an IRI string looks like a Knora role IRI.
```
Comment (irinaschubert) [inline dsp-shared/src/main/scala/dsp/valueobjects/V2.scala:126]:
```suggestion
   * Returns `true` if an IRI string looks like a role IRI.
```

Comment (irinaschubert) [reply dsp-shared/src/main/scala/dsp/valueobjects/V2.scala:126]: The term "Knora" is deprecated and shouldn't be used anymore.

File: dsp-shared/src/main/scala/dsp/valueobjects/Role.scala [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  sealed abstract case class LangString private (value: String, isoCode: String)
```
Comment (irinaschubert) [inline dsp-shared/src/main/scala/dsp/valueobjects/Role.scala:19]: I think it would be more intuititve to turn the paramters' order around: have the isoCode and then the value. This would be the same order as the value is represented when given by or returned to the user, e.g. {"en": "some value"}.

**Review by irinaschubert — APPROVED** [review]
LGTM, I mostly mentioned things that could be cleaned up

File: dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala:11]: This is referring to the naming I was also pointing out at some point, just named the file wrong. Thanks for spotting.

File: dsp-role/core/src/main/scala/dsp/role/api/RoleRepo.scala [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+@accessible
+trait RoleRepo {
+
+  /**
+   * Writes a role into the repository, while both creating or updating a role.
+   *
+   * @param r the role to write
+   * @return Unit
+   */
+  def storeRole(r: Role): UIO[RoleId]
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/api/RoleRepo.scala:27]: In the case then the parameter and its type are the same and it's strongly typed, it could be named: r, role, value, whatever and one will be always notified that's the `Role` type needs to be passed into the method.

[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/api/RoleRepo.scala:27]: In strongly typed languages the type of param, if not omitted, speaks for itself. Therefore  in most of the cases instead of repeating the type in the name it could be shortened to first letter of the type, or named: value, param, whatever, because at the end the IDE will inform one need to pass something of the defined (`Role` in this case) type.

File: dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+ * @param name the role's name
+ * @param description the role's description
+ * @param users the role's users
+ * @param permission the role's permission
+ */
+sealed abstract case class Role private (
+  id: RoleId,
+  name: LangString,
+  description: LangString,
+  users: List[RoleUser], // List[User]
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala:32]: This is actually the question how `User` entity should be represented in here. AFAIR we talked about it and it doesn't have to be equal to the `User` entity already defined. Therefore I've simplify it here and it has only ID inside, and to distinguish the names I've used `RoieUser`. This should be discussed while finalizing the `Role` domain model.

File: dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+  def compare(that: Role): Int = self.id.iri.toString().compareTo(that.id.iri.toString())
+  //  Boolean = self.id.equals(that.id)
+
+  /**
+   * Updates the role's name.
+   *
+   * @param newValue new role's name to update
+   * @return updated [[Role]]
+   */
+  def updateName(newValue: LangString): Validation[BadRequestException, Role] =
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala:51]: In opposite to that `r: Role` here maybe the param name could be name/newName, because type doesn't say what it is, but:
-  `def updateName(newValue: LangString)` is very clear the method will update the Name and requires its **new value** in `LangString` type to be passed
- we have the docString which clarify it even more

File: dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+      self.permission
+    )
+
+  /**
+   * Updates the role's description.
+   *
+   * @param newValue new role's description to update
+   * @return updated [[Role]]
+   */
+  def updateDescription(newValue: LangString): Validation[BadRequestException, Role] =
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala:66]: as aobve

File: dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+      self.permission
+    )
+
+  /**
+   * Updates the role's users.
+   *
+   * @param newValue new role's users to update
+   * @return updated [[Role]]
+   */
+  def updateUsers(newValue: List[RoleUser]): Validation[BadRequestException, Role] =
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala:81]: as above

File: dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+      self.permission
+    )
+
+  /**
+   * Updates the role's permission.
+   *
+   * @param newValue new role's permission to update
+   * @return updated [[Role]]
+   */
+  def updatePermission(newValue: Permission): Validation[BadRequestException, Role] =
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/domain/RoleDomain.scala:96]: as above

File: dsp-role/core/src/test/scala/dsp/role/domain/RoleDomainSpec.scala [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+
+  private val createRoleTest = suite("createRole")(
+    test("create a role") {
+      (
+        for {
+          id          <- RoleTestData.id1
+          name        <- RoleTestData.name1
+          description <- RoleTestData.description1
+          users        = RoleTestData.users1
+          permission  <- RoleTestData.permission1
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/test/scala/dsp/role/domain/RoleDomainSpec.scala:38]: I think I kind of similarly commented on the user slice, but was left like that so here it is like it is there.

File: dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala:19]: True, but I didn't want to create value object for it simply because I don't know how it's going to looks like - if this is going to be complete `User` entity or something else. This should be delegate to finisher task, once we define the `Role` completely.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+   * Retrieves all roles (sorted by IRI).
+   */
+  def getRoles(): UIO[List[Role]] =
+    repo
+      .getRoles()
+      .map(_.sorted)
+      .tap(_ => ZIO.logInfo("Retrieved all roles."))
+
+  /**
+   * Retrieves the role by ID.
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:36]: Not an English expert here, but this is about the specific and defined Role which is known and get by its ID.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+final case class RoleHandler(repo: RoleRepo) {
+
+  /**
+   * Retrieves all roles (sorted by IRI).
+   */
+  def getRoles(): UIO[List[Role]] =
+    repo
+      .getRoles()
+      .map(_.sorted)
+      .tap(_ => ZIO.logInfo("Retrieved all roles."))
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:33]: Good point. I will leave it as it is, but add this topic to our list.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+      role        <- getRoleById(id)
+      updatedRole <- role.updateName(newValue).toZIO
+      _           <- repo.storeRole(updatedRole)
+    } yield id).tap(_ => ZIO.logInfo(s"Updated name with new value: ${newValue.value}"))
+
+  /**
+   * Updates the description of the role.
+   *
+   *  @param id  the roles's ID
+   *  @param newValue  the new description value
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:85]: It doesn't work automatically for me and I really don't want to waste time for do that manually. Check the user implementation, it looks like the same. This might relat to some formatting changes we made last time.

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala [hunk 30]
```diff
@@ -0,0 +1,170 @@
...
+        updatedRoleId <- handler.updateDescription(roleId, newValue)
+        retrievedRole <- handler.getRoleById(updatedRoleId)
+      } yield assertTrue(retrievedRole.name == name1) &&
+        assertTrue(retrievedRole.description == newValue) &&
+        assertTrue(retrievedRole.users == users1) &&
+        assertTrue(retrievedRole.permission == permission1)
+    },
+    // test("store a role and update its users") {
+    //   // implment adding users to the role
+    // },
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:118]: TBH I don't know how this should work and it's related to not finalized `Role` domain model. I'll remove it from here, but will leave questions related to the domain model somewhere else (not sure if here in the code or maybe Figma is better)?

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala (hunk 4 above)
Comment (mpro7) [reply dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala:35]: In this construct the suites are defined on the layer level - see bottom vals. At the end it doesn't bring much, only indentions, and one more level for only one of the layer suites, which is strange. Also in both ways the `+ RoleRepoLive` is not printed out...

```
    + store two roles and retrieve them all
    + store and delete a role
    + store a role and get it by ID
+ RoleRepoMock
  + roleTests
    + store and delete a role
    + store a role and get it by ID
    + store two roles and retrieve them all
 ```

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+      } +
+      test("store and delete a role") {
+        for {
+          role1           <- role1.toZIO
+          roleId          <- RoleRepo.storeRole(role1)
+          idOfDeletedRole <- RoleRepo.deleteRole(roleId)
+          isIdDeleted     <- RoleRepo.getRoleById(idOfDeletedRole).exit
+        } yield assertTrue(roleId == role1.id) &&
+          assert(isIdDeleted)(fails(equalTo(None)))
+      }
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala:51]: Yes, as I said I did only basic things following what was done on the user slice. This would be improved in the successor task, once we finalize the `Role` entity.

File: dsp-shared/src/main/scala/dsp/valueobjects/Id.scala (hunk 6 above)
Comment (mpro7) [reply dsp-shared/src/main/scala/dsp/valueobjects/Id.scala:37]: Ok

File: dsp-shared/src/main/scala/dsp/valueobjects/Iri.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-shared/src/main/scala/dsp/valueobjects/Iri.scala:199]: Agree! Sounds like a task in one of the initiatives.

File: dsp-shared/src/main/scala/dsp/valueobjects/Role.scala [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+  val Modify: String = "modify"
+  val Delete: String = "delete"
+  val Admin: String  = "admin"
+  val availablePermissions: Set[String] = Set(
+    View,
+    Create,
+    Modify,
+    Delete,
+    Admin
+  )
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-shared/src/main/scala/dsp/valueobjects/Role.scala:68]: What's wrong with snipped above?

File: dsp-shared/src/test/scala/dsp/valueobjects/IriSpec.scala [hunk 33]
```diff
@@ -21,6 +21,8 @@ object IriSpec extends ZIOSpecDefault {
   val listIriWithUUIDVersion3    = "http://rdfh.ch/lists/0803/6_xROK_UN1S2ZVNSzLlSXQ"
   val validProjectIri            = "http://rdfh.ch/projects/0001"
   val projectIriWithUUIDVersion3 = "http://rdfh.ch/projects/tZjZhGSZMeCLA5VeUmwAmg"
+  val validRoleIri               = "http://rdfh.ch/roles/jDEEitJESRi3pDaDjjQ1WQ"
+  val roleIriWithUUIDVersion3    = "http://rdfh.ch/roles/cCmdcpn2MO211YYOplR1hQ"
   val validUserIri               = "http://rdfh.ch/users/jDEEitJESRi3pDaDjjQ1WQ"
   val userIriWithUUIDVersion3    = "http://rdfh.ch/users/cCmdcpn2MO211YYOplR1hQ"
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-shared/src/test/scala/dsp/valueobjects/IriSpec.scala:27]: If this is used only in test and these are not interact each other, I don't see there should be a problem with that. But changed these.

File: dsp-role/core/src/main/scala/dsp/role/api/RoleRepo.scala [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+ * Needs to be used by the role repository implementations.
+ */
+@accessible
+trait RoleRepo {
+
+  /**
+   * Writes a role into the repository, while both creating or updating a role.
+   *
+   * @param r the role to write
+   * @return Unit
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-role/core/src/main/scala/dsp/role/api/RoleRepo.scala:25]: Not exactly copy pasting here, but yeah, the returned type is `UIO[RoleId]`

File: dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala (hunk 7 above)
Comment (mpro7) [reply dsp-role/core/src/test/scala/dsp/role/api/RoleRepoSpec.scala:11]: Well this I'm not sure. You also have it in the user slice.

File: dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala (hunk 9 above)
Comment (mpro7) [reply dsp-role/core/src/test/scala/dsp/role/sharedtestdata/RoleTestData.scala:19]: IMO in this basic setup most important is to have non-empty list and if it is one, two or more users, it doesn't matter. Especially that the `Role` entity is not finalized, so the user property isn't clear how it's going to be like.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala (hunk 10 above)
Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:27]: True, but I've noticed we do not always have it there (also in the new code - check user slice). We should agree on the details level and keep it the same everywhere.

Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:27]: Added `@return` for every method in this file.

File: dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala (hunk 11 above)
Comment (mpro7) [reply dsp-role/handler/src/main/scala/dsp/role/handler/RoleHandler.scala:74]: I'm not sure if I get that comment. Param is always explained by its type and we also have it described in docstring. If the type would change to `Seq[LangString]` the `newValue` name for the param is still valid :) At the end it's still the same name translated to languages? We had this conversation before, so we should stick in this cases to singular. Or change the way we deal with internationalization.

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala (hunk 13 above)
Comment (mpro7) [reply dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:28]: Updated.

File: dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala (hunk 14 above)
Comment (mpro7) [reply dsp-role/handler/src/test/scala/dsp/role/handler/RoleHandlerSpec.scala:170]: I'm not sure about the users, this was not yet clarified in the `Role` entity how will that work, but should the deletion of the description be possible if its required? Added missing test for delete method.

File: dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala (hunk 15 above)
Comment (mpro7) [reply dsp-role/repo/src/test/scala/dsp/role/repo/impl/RoleRepoImplSpec.scala:54]: Sorry I don't get this comment.

File: dsp-shared/src/main/scala/dsp/valueobjects/Role.scala (hunk 19 above)
Comment (mpro7) [reply dsp-shared/src/main/scala/dsp/valueobjects/Role.scala:19]: Subject to discuss.
